lib/ansible/modules/cloud/amazon/ec2_ami.py

```diff
@@ -542,69 +535,102 @@ def deregister_image(module, connection):
-                group_names = launch_permissions.get('group_names')[0] if launch_permissions.get('group_names') else None
-                user_ids = launch_permissions.get('user_ids')[0] if launch_permissions.get('user_ids') else None
-                launch_perms_add = {'Add': [{}]}
-                if group_names:
-                    launch_perms_add['Add'][0]['Group'] = group_names
-                if user_ids:
-                    launch_perms_add['Add'][0]['UserId'] = user_ids
-                image.modify_attribute(Attribute='launchPermission', LaunchPermission=launch_perms_add)
-            elif set_permissions and set_permissions[0].get('UserId') is not None and set_permissions[0].get('Group') is not None:
-                image.modify_attribute(
-                    Attribute='launchPermission',
-                    LaunchPermission={
-                        'Remove': [{
-                            'Group': (set_permissions.get('Group') or ''),
-                            'UserId': (set_permissions.get('UserId') or '')
-                        }]
-                    })
-            else:
-                module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False,
-                                 **get_ami_info(get_image_by_id(module, connection, image_id)))
-            module.exit_json(msg="AMI launch permissions updated.", launch_permissions=launch_permissions, set_perms=set_permissions, changed=True,
-                             **get_ami_info(get_image_by_id(module, connection, image_id)))
-        else:
-            module.exit_json(msg="AMI not updated.", launch_permissions=set_permissions, changed=False,
-                             **get_ami_info(get_image_by_id(module, connection, image_id)))
-    except botocore.exceptions.ClientError as e:
-        module.fail_json(msg="Error updating image - " + str(e), exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+        to_add = [dict(Group=group) for group in to_add_groups] + [dict(UserId=user_id) for user_id in to_add_users]
+        to_remove = [dict(Group=group) for group in to_remove_groups] + [dict(UserId=user_id) for user_id in to_remove_users]
+
+        if to_add or to_remove:
+            try:
+                connection.modify_image_attribute(ImageId=image_id, Attribute='launchPermission',
+                                                  LaunchPermission=dict(Add=to_add, Remove=to_remove))
+                changed = True
+            except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+                module.fail_json_aws(e, msg="Error updating launch permissions")
+
+    desired_tags = module.params.get('tags')
+    current_tags = boto3_tag_list_to_ansible_dict(image.get('Tags'))
+    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))
```

Comment:
This should probably check if tags is not None first.
If the user doesn't specify tags in their play:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py", line 699, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py", line 692, in main
    update_image(module, connection, module.params.get('image_id'))
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py", line 578, in update_image
    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_modlib.zip/ansible/module_utils/ec2.py", line 757, in compare_aws_tags
TypeError: argument of type 'NoneType' is not iterable

fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py\", line 699, in <module>\n    main()\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py\", line 692, in main\n    update_image(module, connection, module.params.get('image_id'))\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py\", line 578, in update_image\n    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_modlib.zip/ansible/module_utils/ec2.py\", line 757, in compare_aws_tags\nTypeError: argument of type 'NoneType' is not iterable\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
```

Reply:
That's a good spot - you also made me realise that if setting `purge_tags` and not setting `tags` at all, it could remove all existing tags.